### PR TITLE
feat: add optional function split loop

### PR DESCRIPTION
**Pull request by RossWilliams**
Optional function transform-split-loop can break down an EDI file at the sub-transaction level to make processing large data easier.

The optional function is not deployed by default and has its own deployment script to encapsulate the function and allow easy modification by users for their own need.

fix: detecting error message for creating transaction settings
chore: refactor to allow re-use of function deploy code

### CODE REVIEW

**Review by rsioss — APPROVED** [review]
left several (primarily nit) comments for consideration. definitely good to 🚢  as-is, and any comments you agree with can be follow-ups if you'd like 🚀

File: src/optional-functions/transform/split-loop/deploy.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import dotenv from "dotenv";
+import {
+  deployEventBinding,
+  deployFunctionAtPath,
+} from "../../../lib/functions.js";
+
+// change to root directory
+process.chdir(new URL("../../../..", import.meta.url).pathname);
```
[the remark this answers is not in the document]
Comment (RossWilliams) [reply src/optional-functions/transform/split-loop/deploy.ts:8]: agree, was talking with Brian today about a function template config file to specify these things. I'm pretty slow at design, so I'll need to think about it for a bit.

File: src/optional-functions/transform/split-loop/edi-splitter.ts [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+  stControlNumbers: string[];
+}
+
+// split edi file by transaction set preserving ISA and GS headers
+// ISA -- GS -- ST850 -- SE -- ST850 -- SE -- ST977 -- SE -- GE -- IEA split into array
+// { "850", ISA -- GS -- ST850 -- GE -- IEA },
+// { "850", ISA -- GS -- ST850 -- GE -- IEA },
+// { "977", ISA -- GS -- ST977 -- GE -- IEA }
+
+export const ediSplitter = (
```
[the remark this answers is not in the document]
Comment (RossWilliams) [reply src/optional-functions/transform/split-loop/edi-splitter.ts:19]: This version has a few fixes in it, the original had some subtle bugs (very not fun). Happy for you to switch it out, but notice the sub-transaction split code.

File: src/optional-functions/transform/split-loop/configuration.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { Configuration } from "./types";
+
+const coreIngestionBucket = process.env.CORE_INGESTION_BUCKET_NAME!;
+
+export default {
+  configuration: {
```
[the remark this answers is not in the document]
Comment (RossWilliams) [reply src/optional-functions/transform/split-loop/configuration.ts:6]: Time constraints, good idea for the future. I started writing a stash config, but there are subtle issues that made it more difficult than I had time to solve.

File: src/optional-functions/transform/split-loop/handler.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    const ediString = await ediObject.body.transformToString();
+    const splitEdi = ediSplitter(
+      ediString,
+      config.configuration.splitSegment,
+      config.configuration.chunkSize,
+      config.configuration.transactionSetId
+    );
+
+    const { errors } = await PromisePool.for(splitEdi)
+      .withConcurrency(5)
```
[the remark this answers is not in the document]
Comment (RossWilliams) [reply src/optional-functions/transform/split-loop/handler.ts:54]: no, this is set at a level that will not cause database limits to be hit.

File: src/optional-functions/transform/split-loop/types.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+export interface Configuration {
```
[the remark this answers is not in the document]
Comment (RossWilliams) [reply src/optional-functions/transform/split-loop/types.ts:1]: once we move to stash, yes.

File: src/optional-functions/transform/split-loop/edi-splitter.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rsioss) [reply src/optional-functions/transform/split-loop/edi-splitter.ts:19]: Right, the loop splitting code I was referring to above that is in my local branch was for the sub transaction splitting use case (for 834s, but customizable for the loop segments as you've done here). If you have test cases for the previous bugs we can include those to make sure they are covered if/when we migrate to the other approach.

Your point regarding the bugs is exactly why I wasn't a fan of the regex approach. Much harder to trace, debug, and reason about as opposed to the implementation linked above

[the remark this answers is not in the document]
Comment (RossWilliams) [reply src/optional-functions/transform/split-loop/edi-splitter.ts:19]: I have a test case for this file that has multiple GS segments and breaks down a transaction set by the LX loop. Reading your code link, it is a lot easier to read the breakdown into the transaction level.

**Pull request merged**